tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmDictOperationInterpreter.kt

```diff
@@ -1669,15 +1998,76 @@ class TvmDictOperationInterpreter(
         val dictId = DictId(keyLength)
         val keySort = ctx.mkBvSort(keyLength.toUInt())
 
-        val resultElement = scope.calcOnState { makeSymbolicPrimitive(keySort) }
-        val resultElementExtended = extendDictKey(resultElement, keyType)
-
         // since these entries were stored during execution, value overflow constraints have already been asserted
         val allSetEntries =
-            scope.calcOnState {
+            scope.calcOnStateCtx {
                 memory.setEntries(dictCellRef, dictId, keySort, DictKeyInfo)
             }
+        val isInput =
+            allSetEntries.isInput || scope.calcOnState { inputDictionaryStorage.memory.keys.contains(dictCellRef) }
+        if (isInput) {
+            val dictOriginalConcrete =
+                dictCellRef as? UConcreteHeapRef
+                    ?: TODO("ite refs are not supported yet")
+            val (inputDict, rootInputDictInfo) =
+                scope.calcOnState {
+                    readInputDictionary(dictOriginalConcrete, keySort, keyKind)
+                }
+            val keyResultSymbol = scope.calcOnState { makeFreshKeyConstant(keySort, keyKind) }
+            val freshConstantForInput = scope.calcOnState { makeFreshKeyConstant(keySort, keyKind) }
+            val dictMaxResult =
+                inputDict.doDictMaxMin(
+                    ctx,
+                    rootInputDictInfo,
+                    mode == DictMinMaxMode.MAX,
+                    freshConstantForInput,
+                    keyResultSymbol,
+                    keyKind == DictKeyKind.SIGNED_INT,
+                )
+            scope.assert(ctx.mkAnd(dictMaxResult.constraintSet))
+                ?: return
+
+            scope.calcOnState {
+                inputDictionaryStorage =
+                    inputDictionaryStorage.set(
+                        dictOriginalConcrete,
+                        inputDict,
+                        dictMaxResult.newInputDictRootInformation,
+                    )
+            }
+
+            val value = scope.calcOnState { dictGetValue(dictCellRef, dictId, keyResultSymbol.expr) }
+            val unwrappedValue =
+                unwrapDictValue(scope, value, valueType)
+                    ?: return
+
+            assertDictValueDoesNotOverflow(scope, dictId, value)
+                ?: return
```

Comment:
Cannot we make this assertion with your universal quantifiers? Then, we probably won't even need this function.